fuel-vm/src/interpreter/flow.rs

```diff
@@ -149,138 +161,385 @@ where
+        <S as StorageInspect<ContractsAssets>>::Error: Into<std::io::Error>,
+    {
+        let call = self.memory.call_params.try_from(self.memory.memory)?;
+        let asset_id = self.memory.asset_id.try_from(self.memory.memory)?;
+
+        let mut frame = call_frame(self.registers.copy_registers(), &self.storage, call, asset_id)?;
+
+        let profiler = ProfileGas {
+            pc: self.registers.read_registers.pc.as_ref(),
+            is: self.registers.read_registers.is.as_ref(),
+            current_contract: self.current_contract,
+            profiler: self.profiler,
+        };
+        dependent_gas_charge(
+            self.registers.read_registers.cgas.as_mut(),
+            self.registers.read_registers.ggas.as_mut(),
+            profiler,
+            self.gas_cost,
+            frame.total_code_size(),
+        )?;
+
+        if let Some(source_contract) = self.current_contract {
+            balance_decrease(
+                self.storage,
+                &source_contract,
+                frame.asset_id(),
+                self.params.amount_of_coins_to_forward,
+            )?;
         } else {
-            let source_contract = *self.internal_contract()?;
-            self.balance_decrease(&source_contract, &asset_id, b)?;
+            let amount = self.params.amount_of_coins_to_forward;
+            external_asset_id_balance_sub(self.runtime_balances, self.memory.memory, frame.asset_id(), amount)?;
         }
 
-        if !self
-            .transaction()
-            .input_contracts()
-            .any(|contract| call.to() == contract)
-        {
-            self.panic_context = PanicContext::ContractId(*call.to());
+        if !self.input_contracts.iter().any(|contract| call.to() == contract) {
+            *self.panic_context = PanicContext::ContractId(*call.to());
             return Err(PanicReason::ContractNotInInputs.into());
         }
 
         // credit contract asset_id balance
-        self.balance_increase(call.to(), &asset_id, b)?;
-
-        let forward_gas_amount = cmp::min(self.registers[RegId::CGAS], d);
+        balance_increase(
+            self.storage,
+            call.to(),
+            &asset_id,
+            self.params.amount_of_coins_to_forward,
+        )?;
+
+        let forward_gas_amount = cmp::min(
+            *self.registers.read_registers.cgas,
+            self.params.amount_of_gas_to_forward,
+        );
 
         // subtract gas
-        self.registers[RegId::CGAS] = arith::sub_word(self.registers[RegId::CGAS], forward_gas_amount)?;
+        *self.registers.read_registers.cgas = arith::sub_word(*self.registers.read_registers.cgas, forward_gas_amount)?;
 
-        *frame.set_context_gas() = self.registers[RegId::CGAS];
-        *frame.set_global_gas() = self.registers[RegId::GGAS];
+        *frame.set_context_gas() = *self.registers.read_registers.cgas;
+        *frame.set_global_gas() = *self.registers.read_registers.ggas;
```

Comment:
[nit] this doesn't look like a very conventional way to name rust apis, usually setters would take the value as input, while getting mutable access to some field is done with a `*_mut` method. 

```suggestion
        *frame.context_gas_mut() = *self.registers.read_registers.cgas;
        *frame.global_gas_mut() = *self.registers.read_registers.ggas;
```

Reply:
I agree, it would be more in "rust wat"